augur/io/metadata.py

```diff
@@ -455,6 +455,88 @@ def write_records_to_tsv(records, output_file):
             tsv_writer.writerow(record)
 
 
+class Metadata:
+    """Represents a metadata file."""
+
+    path: str
+    """Path to the file on disk."""
+
+    dialect: csv.Dialect
+    """Inferred dialect of metadata."""
+
+    columns: Sequence[str]
+    """Columns extracted from the first row in the metadata file."""
+
+    id_column: str
+    """Inferred ID column."""
+
+    def __init__(self, path: str, delimiters: Sequence[str], id_columns: Sequence[str]):
+        """
+        Parameters
+        ----------
+        path
+            Path of the metadata file.
+        delimiters
+            Possible delimiters to use, in order of precedence.
+        id_columns
+            Possible ID columns to use, in order of precedence.
+        """
+        self.path = path
+
+        # Infer the dialect.
+        self.dialect = _get_dialect(self.path, delimiters)
+
+        # Infer the column names.
+        with self.open() as f:
+            reader = csv.reader(f, delimiter=self.dialect.delimiter)
+            try:
+                self.columns = next(reader)
+            except StopIteration:
+                raise AugurError(f"{self.path}: Expected a header row but it is empty.")
+
+        # Infer the ID column.
+        self.id_column = self._find_first(id_columns)
+
+    def open(self, **kwargs):
+        """Open the file with auto-compression/decompression."""
+        return open_file(self.path, **kwargs)
+
+    def _find_first(self, columns: Sequence[str]):
+        """Return the first column in `columns` that is present in the metadata.
+        """
+        for column in columns:
+            if column in self.columns:
+                return column
+        raise AugurError(f"{self.path}: None of ({columns!r}) are in the columns {tuple(self.columns)!r}.")
+
+    def rows(self, skip_blank_lines: bool = True, strict: bool = True):
+        """Yield rows in a dictionary format.
+
+        Parameters
+        ----------
+        skip_blank_lines
+            If True, skip over blank lines.
+        strict
+            If True, raise an error when a row contains more or less than the number of expected columns.
+        """
+        with self.open() as f:
+            reader = csv.DictReader(f, delimiter=self.dialect.delimiter, fieldnames=self.columns, restkey=None, restval=None)
+
+            # Skip the header row.
+            next(reader)
+
+            for row in reader:
+                if skip_blank_lines and not row:
+                    continue
```

Comment:
Good catch, I'll see if the condition can be changed to something more explicit than `not row`.

Reply:
It'll be tricky to allow blank lines since the skipping is [done by DictReader](https://github.com/python/cpython/blob/main/Lib/csv.py#L181-L185) and can't be disabled directly ([StackOverflow](https://stackoverflow.com/a/27707768)). Practically, I don't see a use case for keeping them, so I'll just remove this option.